### ISSUE TITLE
Feature request: `utf-16le-bom` & `utf-16be-bom` charsets

### ISSUE THREAD

**Issue by mataha** [editorconfig/editorconfig#467]
Certain times I have a dubious pleasure of working with UTF-16 files that need to have a BOM in order to be read correctly - usual suspects involve Windows Registry files and some PowerShell scripts. When ceding control over character sets to editorconfig I become unable to preserve proper encoding in those files.

Ideally, I would like to include something like this in my `.editorconfig`:

```ini
[*.{dat,reg}]
charset = utf-16le-bom
end_of_line = crlf
```

There are tools that respect such files, e.g. git via `.gitattributes`:

```gitattributes
*.reg text working-tree-encoding=UTF-16LE-BOM eol=crlf
```

**Comment by cxw42** [comment]
@mataha Do you still need this change?  Have you hacked your EC plugin to support these values anyway? :)

**Comment by greut** [comment]
From what I've seen, it's usually `utf-16le` and `utf-16be` that are in use. The BOM is not mandatory but feels like it. Those values are supported by some plugins already.

**Comment by mataha** [comment]
> @mataha Do you still need this change? Have you hacked your EC plugin to support these values anyway? :)

I still do need this.

My temporary solution involves `charset = unset` - which is less than ideal, but the only one that works in shared repositories.

**Comment by cxw42** [comment]
Thanks!  I personally have no objections.

**Comment by xuhdev** [comment]
I also see this as appropriate. As long as we don't make `charset` a huge map of all encodings, we should be able to have decent editor coverage.

**Comment by greut** [comment]
According to https://en.wikipedia.org/wiki/Byte_order_mark#UTF-32, the same would be needed for `utf-32le` and `utf-32le` or am I missing something?

**Comment by xuhdev** [comment]
I agree, if we add bom we should add them altogether.

**Comment by stevewgr** [comment]
Strange that it is even a "feature request" ticket. This should exists, since in some cases utf-16 are treated as binary for some tools, like git where we have to mark them with gitattributes:
```
*.rc text working-tree-encoding=UTF-16LE-BOM eol=LF
*.rc2 text working-tree-encoding=UTF-16LE-BOM eol=LF
```

**Comment by tats-u** [comment]
All Windows ANSI (and OEM) encodings including non-Latin ones like Shift_JIS are also covered by this issue. The title must be changed.

Visual Studio (mainly C++ or 2022 or prior) occasionally forces users to use such encodings even today.

**Comment by tats-u** [comment]
In CJK languages (and a few other non-Euroamerican languages) a single encoding (e.g. Shift_JIS for Japanese) is assigned to both ANSI and OEM encodings. Does someone know when OEM encodings are used as _file_ endings?

**Comment by florianb** [comment]
@cxw42 @xuhdev

Would it make sense to simply add the support of the `iconv -l` based list of encodings to the specs, leaving it open to the maintainers which encoding to support?

List of current Encodings:
```
  437, 500, 500V1, 850, 851, 852, 855, 856, 857, 858, 860, 861, 862, 863, 864,
  ...snip...
  WINDOWS-1254, WINDOWS-1255, WINDOWS-1256, WINDOWS-1257, WINDOWS-1258,
  WINSAMI2, WS2, YU
```

**edit:** shortened the list as the IANA-specs should be used.

**Comment by tats-u** [comment]
IANA-registered names are more standardized.

https://www.iana.org/assignments/character-sets/character-sets.xhtml

Over 95% of source code non-Unicode encodings we encounter today are either ANSI or OEM code pages for some language on Windows. Adding historically used ISO-2022-based 7-bit encodings for email would cover well over 99%. Supporting EBCDIC-based encodings—which are incompatible with plain ASCII—can be made optional. Also, most of the source code encodings claiming to be Latin-1 that we encounter are actually Windows-1252.

**Comment by tats-u** [comment]
The WHATWG Encoding Standard may also be useful:

https://encoding.spec.whatwg.org/#names-and-labels

**Comment by florianb** [comment]
Good point @tats-u - thank you very much! The IANA standard is indeed the specification for iconv as well.

**Comment by tats-u** [comment]
In some encodings, Windows' extended versions may be widely distributed under the original names. For example, when Shift_JIS is specified, it almost invariably refers to the Windows extension Windows-31J. Windows extended superset should be referred especially in Windows and Linux (i.e. other than macOS) in such cases.

**Comment by florianb** [comment]
I understand, so you'd propose to merge the iana and the whatwg.org-encoding-spec?

That's no issue for me, i'd add a proposal referring to the specs and creating an example mapping for the plugin-devs. At the end it's to the plugin devs to map it to the corresponding labesl of the target environment.

**Comment by tats-u** [comment]
> I understand, so you'd propose to merge the iana and the whatwg.org-encoding-spec?

Yes, and I forgot MacJapanese (Shift_JIS variant for macOS, not fully compatible with Windows' Windows-31J) has not been registered by IANA or WHATWG.

If WHATWG maps a particular label to a superset of the encoding defined by IANA, we should prioritize the WHATWG mapping.

In addition to this, some IANA labels starting with "cs" are redundant. Are they necessary?

**Comment by florianb** [comment]
In my opinion the main service of mapping any given values in that property should remain to the plugin-dev.

As soon as we start to refine the standards we must take over the responsibility to solve questions like this. And i don't think the EditorConfig group should do so.

So then the WHATWG-mapping should be priority. Thankfully they also specify the algorithm how to map the label to the given encoding.

**Comment by cxw42** [comment]
I would be happy with either IANA or WHATWG. 

It looks to me IANA defines more encodings than WHATWG, so if we wanted the broadest possible support I would recommend IANA.

**Comment by tats-u** [comment]
https://github.com/unicode-org/icu/blob/main/icu4c/source/data/mappings/convrtrs.txt

I found this in ICU seems to be a superset of the IANA list. I don't think it is well organized though.

**Comment by cxw42** [comment]
@tats-u thanks!  I see that ICU file mentions a rendered list at https://icu4c-demos.unicode.org/icu-bin/convexp .  I took a quick look, and that page lists:

> [UTF16_PlatformEndian](https://icu4c-demos.unicode.org/icu-bin/convexp?conv=UTF16_PlatformEndian&s=ALL)	UTF16_PlatformEndian
> [UTF16_OppositeEndian](https://icu4c-demos.unicode.org/icu-bin/convexp?conv=UTF16_OppositeEndian&s=ALL)	UTF16_OppositeEndian

Those two by definition are not useful for EditorConfig since EC is platform-independent.  Therefore, would recommend we not use that list.

**Comment by cxw42** [comment]
### Discussion

Looking back at the original request, neither IANA nor WHATWG provides both utf16be-bom and utf16le-bom.  Therefore, neither alone provides the requested feature :( .
* WHATWG expressly prohibits serializing UTF-16: https://encoding.spec.whatwg.org/#output-encodings
  > To get an output encoding ... If encoding is ... UTF-16BE/LE, then return UTF-8
* IANA's UTF16 points to [RFC2781](https://www.iana.org/go/rfc2781), which [says](https://datatracker.ietf.org/doc/html/rfc2781#section-4.3):
  > Text labelled with the "UTF-16" charset might be serialized in either big-endian or little-endian order.
* Git's [`working-tree-encoding` attribute](https://git-scm.com/docs/gitattributes/2.18.0#_working_tree_encoding), mentioned by the OP, uses `iconv` names.  As I mentioned in my previous post, iconv includes some entries that are not platform-independent, so I don't think we should use `iconv`.

Therefore, how about this?---

### Proposal

* EC `charset` values are: `(<iana> ('&' <modifier> | '!' <modifier>)*) | 'EUC-JP' | 'latin1' | 'utf-8-bom'` (the latter three being legacy values)
  * `&<modifier>` means "including `<modifier>`".  `!<modifier>` means "excluding `<modifier>`".  For example, plugins MUST NOT add a BOM if `charset` includes `!bom`.
  * `charset` values are case-insensitive (as in the current spec).
  * `&` and `!` do not occur in either WHATWG or IANA names, so are available for our use.
* The list of allowed `<modifier>` values is specified by EC.
* Initially, the only allowed `<modifier>` value is `bom`.
* `utf-8`, `utf-16`, `utf-16be`, and `utf-16le` imply `!bom` unless `&bom` is specified.  This is for simplicity.  (See discussion below.)
* `<iana>` is a value from the "name" column of <https://www.iana.org/assignments/character-sets/character-sets.xhtml>.
* The legacy values mean:
  * `EUC-JP` == `Extended_UNIX_Code_Packed_Format_for_Japanese`
    * This is the only one we need from the IANA "preferred mime name" column, so I called it out separately.
  * `latin1` == `ISO-8859-1`
  * `utf-8-bom` == `utf-8&bom`

The OP's use case would be `charset = utf-16le&bom`.

As in the spec currently, plugins ignore invalid/unsupported values.  Therefore, if a plugin understands `utf-8` but not `utf-8&bom`, it should ignore any `charset = utf-8&bom`.

#### UTF-* and default-`!bom`

I am proposing defaulting to `!bom` to keep things simple.  This is in keeping with the current spec, which expressly discourages utf-8-bom, for example.

However, with regards to UTF-16, this is a difference from [RFC2781 sec. 3.3](https://datatracker.ietf.org/doc/html/rfc2781#section-3.3).  If we went by the RFC:
* `utf-16` would imply `&bom`.  The RFC says 'formats that mandate a BOM in UTF-16 text ... \[require] the use of the "UTF-16" tag only.'  Therefore `utf-16` == `utf-16&bom`.
* `utf-16be` and `utf-16le` would imply `!bom`.  The RFC says "systems labelling UTF-16BE text *\[or UTF-16LE text]* MUST NOT prepend a BOM to the text."

If others think it would be less confusing to stick more closely to the RFC, then we could certainly do so.

### MacJapanese

I'm not sure how to handle this case.  If anyone is actually using MacJapanese, please post here and let us know!  [Wikipedia](https://en.wikipedia.org/wiki/Shift_JIS#MacJapanese) says there are actually several different variants of MacJapanese.

Since MacJapanese is family of Shift_JIS variants, we could represent them with charset modifiers.  E.g., what Wikipedia describes as:
* "Code page 10001" could be `charset = Shift_JIS&cp10001`
* 'a "PostScript" variant of MacJapanese' could be `charset = Shift_JIS&postscript`
